image_builder/image_builder/docker/builder.py

```diff
@@ -116,30 +132,34 @@ def _wait_for_docker_engine_ready():
 
 
 def _gather_job_templates(
-    config: Config,
     plugin_engine: PluginEngine,
-) -> Dict[str, Tuple[str, Path]]:
-    """Return job name -> (base image name, dockerfile template path)"""
-    job_templates: Dict[str, Tuple[str, Path]] = {}
-    racetrack_tag = os.environ.get('DOCKER_TAG', 'latest')
-    # Standard job types
-    jobs_path = project_root() / 'wrappers' / 'docker'
-    for job_path in jobs_path.iterdir():
-        job_name = job_path.name
-
-        base_image = get_fatman_base_image(config.docker_registry, config.docker_registry_namespace, job_name, racetrack_tag)
-        template_path = job_path / 'fatman-template.Dockerfile'
-        assert template_path.is_file(), f'cannot find Dockerfile template for {job_name} language wrapper: {template_path}'
-
-        job_templates[job_name] = (base_image, template_path)
-
-    # Plugin job types
-    docker_registry_prefix = join_paths(config.docker_registry, config.docker_registry_namespace, 'fatman-base')
-    plugin_results = plugin_engine.invoke_plugin_hook(PluginCore.fatman_job_types, docker_registry_prefix)
+) -> Dict[str, Tuple[Path, Path, str]]:
+    """
+    Load job types from plugins.
+    Return job name -> (base image name, dockerfile template path)
+    """
+    job_templates: Dict[str, Tuple[Path, Path, str]] = {}
+
+    plugin_results: List[Dict[str, Tuple[Path, Path]]] = plugin_engine.invoke_plugin_hook(PluginCore.fatman_job_types)
     for plugin_job_types in plugin_results:
-        for job_name, job_data in plugin_job_types.items():
-            base_image, template_path = job_data
-            assert template_path.is_file(), f'cannot find Dockerfile template for {job_name} language wrapper: {template_path}'
-            job_templates[job_name] = (base_image, template_path)
+        if plugin_job_types:
+            for job_full_name, job_data in plugin_job_types.items():
+                base_image_path, template_path = job_data
+                assert base_image_path.is_file(), f'cannot find base image Dockerfile for {job_full_name} language wrapper: {base_image_path}'
+                assert template_path.is_file(), f'cannot find Dockerfile template for {job_full_name} language wrapper: {template_path}'
+                name_parts = job_full_name.split(':')
+                job_name = name_parts[0]
+                job_version = name_parts[1] if len(name_parts) > 1 else 'latest'
+                job_templates[job_name] = (base_image_path, template_path, job_version)
 
     return job_templates
+
+
+def _image_exists_in_registry(image_name: str):
+    try:
+        shell(f'docker manifest inspect --insecure {image_name}')
```

Comment:
Ahh, so that's clever trick to check whether image exists on docker hub - that was the intention here, right?  Didn't know it. If it was only about localhost docker registry, `docker image inspect` would do the trick. One downside of `docker manifest` that it takes about 10 sec to check for image that exists, and 5 for one that doesn't exist. Btw, it's experimental feature in docker, so worth adding comment about it.

Reply:
Yes, it's supposed to work with localhost registry as well as docker hub / github registry or any on-premises Gitlab registry.
It's true that sometimes it can be slow, though it's the matter of 20ms to check if image exists in local registry on my computer. For ghcr.io registry it lasts about 3 seconds.

If you know a better way to do it, I'll appreciate that.